keps/sig-node/2411-cri-container-log-rotation/README.md

```diff
@@ -0,0 +1,301 @@
+
+## Motivation
+
+Container runtimes that communicate with the kubelet via the Container Runtime Interface (CRI) needed a container log management system. The kubelet was already in charge of determining the container log file path and passing that down to the container runtime so that it can write the container logs there. Thus making the kubelet in charge of rotating the container logs allows the kubelet to manage and access the logs directly without having to call the container runtime. An added advantage of this is that logging agents can ingest files directly without any further integrations with the container runtime.
+
+### Goals
+
+- The kubelet assigns the log path for a container and runtime writes the container output to that path
+- The kubelet periodically checks the disk space occupied by the container logs and rotates them if necessary
+- After the logs are rotated, the kubelet sends a signal to the container runtime to re-open the log file
+- The kubelet exposes a consistent log directory structure with metadata so that any logging agent can integrate with it
+
+### Non-Goals
+
+- Shipping the logs directly to a remote storage
+- Supporting container runtimes that run on a different virtual/physical machine from the kubelet
+
+## Proposal
+
+Graduate the CRIContainerLogRotation feature gate from Beta to Stable. The kubelet already decides the container log directory structure and passes that down to the container runtime. The container runtime then writes the container's logs to this location. This makes the kubelet the best candidate to manage the rotation of the container logs as it already know the log directory structure and has access to it. The CRILogRotation feature gate implementation adds a container log manager package, which manages and rotates the container logs. It also adds 2 flags to the kubelet that allows the user to configure the maximum size of each log file and the maximum number of log files to retain. These flags are **--container-log-max-size** and **--container-log-max-files**.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a kubernetes user, I want to use a CRI container runtime and want the container logs to be managed and rotated by the kubelet.
+
+#### Story 2
+
+As a kubernetes user, I want to integrate a logging agent that aggregates the logs to a remote storage without having to further integrate with the CRI container runtime.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- Loss of some logs during log rotation. There is an open issue on this with a suggested fix https://github.com/kubernetes/kubernetes/issues/64760
```

Comment:
If we are intending to fix this issue (which I think we should), this should be moved under Beta->GA criteria.

Reply:
Added it to the graduation criteria as well